docs/labs/ja_oob1.html

```diff
@@ -0,0 +1,262 @@
+<!DOCTYPE html>
+<html lang="ja">
+<head>
+<meta http-equiv="X-UA-Compatible" content="IE=edge">
+<meta name="viewport" content="width=device-width, initial-scale=1">
+<link rel="stylesheet" href="https://best.openssf.org/assets/css/style.css">
+<link rel="stylesheet" href="checker.css">
+<script src="checker.js"></script>
+<script src="oob1.js"></script>
+<link rel="license" href="https://creativecommons.org/licenses/by/4.0/">
+
+<!-- See create_labs.md for how to create your own lab! -->
+
+</head>
+<body>
+<!-- For GitHub Pages formatting: -->
+<div class="container-lg px-3 my-5 markdown-body">
+<h1>ラボ演習 oob1</h1>
+<p>
+これはセキュアなソフトウェア開発に関するラボ演習です。
+ラボの詳細については、<a href="introduction.html" target="_blank">概要</a>をご覧ください。
+
+<p>
+<h2>タスク</h2>
+<p>
+<b>バッファオーバーリードバグの一例として、OpenSSL の "Heartbleed" を単純化した以下のコードを変更してバグを修正してください</b>
+
+<p>
+<h2>背景</h2>
+<p>
+ほとんど全てのプログラミング言語では、もしプログラムがバッファの外側を読み書きしようとした場合にはバッファをリサイズするか、何等かのエラーを出力するかをデフォルトの挙動としています（例：例外を上げる）。
+なぜなら誤ってバッファの外側を読み書きしようとするのは非常によくあることだからです。
+
+<p>
+しかし、C と C++ は事情が異なります。
+C++ は進歩しており、いくらかは安全になっていますが（例：スマートポインタ）、
+C と C++ においてバッファの外側を読み書きするような動作は多くの場合で
+<i>未定義の動作</i>であり、未定義の動作に対しては
+<i>あらゆることが</i>
+なんの防御もなしに発生することになります。
+実際によく起こるのは、なにか他のデータが読み書きされることです。
+<a href="https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2771r0.html"
+>proposals to improve C++ memory safety</a>
+という提案もありますが、現状では多くの C++ 組込みデータ構造（例えば配列）がメモリセーフではないため、ここでは二つの言語を同時に取り扱います。
+
+<p>
+2014年に明らかになった Heartbleed（CVE-2014-0160）は、バッファオーバーリード脆弱性の一例です。
+Heartbleed は OpenSSL の脆弱性です。OpenSSL は暗号プロトコルである Secure Socket Layer (SSL) とその後継である Transport Layer Security (TLS) を C で実装しており、広く使用されているツールキットです。
+Heartbleed は極めて多くの、例えば
+Google, YouTube, Yahoo!, Pinterest, Blogspot, Instagram, Tumblr,
+Reddit, Netflix, Stack Overflow, Slate, GitHub, Yelp, Etsy, the
+U.S. Postal Service (USPS), Blogger, Dropbox, Wikipedia, and the
+Washington Post
+といった有名なウェブサイトに影響を与えました。
+より詳しくは <a href="https://dwheeler.com/essays/heartbleed.html">こちら</a> を参照してください。
+
+
+<p>
+<h2>タスクの詳細</h2>
+<p>
+ここでは OpenSSL の Heartbleed 脆弱性に対する修正を、<tt>dtls1_process_heartbeat</tt> という関数を修正することで再現します。
+
+<p>
+この時点のコードでは、
+<tt>s->s3->rrec.length</tt>
+というデータ構造はバッファの何バイトが利用可能かを表します。
+もし上限サイズを確認しなければ、バッファを超えた読み出しが簡単に発生してしまいます。
+以下のコードでは２か所を変更することで、この問題を修正しましょう。
+
+<p>
+まず初めに、レスポンスの最短の長さである <tt>(1 + 2 + 16)</tt> が
+<tt>s->s3->rrec.length</tt>
+で取得できる長さを超えていればハートビートを送信せずに 0 を返すようにコードを変更します。
+これは、ハートビートを生成するための領域が全くない場合にはハートビートを生成しないようにするための措置です。
+
+<p>
+続いて、ペイロードがある場合のレスポンスの最短の長さである
+<tt>(1 + 2 + payload + 16)</tt>
+が、
+<tt>s->s3->rrec.length</tt>
+で示されるレスポンスをを格納するための領域より長い場合にも、ハートビートを送信せずに 0 を返すようコードを変更します。
```

Comment:
```suggestion
で示されるレスポンスを格納するための領域より長い場合にも、ハートビートを送信せずに 0 を返すようコードを変更します。
```
Typo. (duplication of "を")

Reply:
Nice catch and thank you @shimos.
FIxed it.

Reply:
@Muuhh-CTJ - if all is fixed, would you please click on "Resolve conversation" in the GitHub interface? That way we know that you believe these have been resolved.

Reply:
OK, thanks David!